### DISCUSSION TITLE
Question related to: window.is_tree,  command.refresh, reg.set and reg.delete

### DISCUSSION THREAD

**Discussion by RubicBG** [moudey/Shell#249] [Q&A]
I have a few questions:

There is something "weird" in the Navigation Pane of Windows Explorer. There are different context menus:
- the 'first menu' is when selecting an object (sel.count=1) (everything is fine here)
![image](https://user-images.githubusercontent.com/127151533/236616123-4c13d621-8535-43ee-8f26-8558e2789558.png)
- the 'second menu' is when choosing between objects (sel.count=1)
![image](https://user-images.githubusercontent.com/127151533/236615853-bb6cad59-8e8d-449f-aa86-eb26c4a22e7f.png)
- the 'third menu' is when no object is selected (sel.count=0)
![image](https://user-images.githubusercontent.com/127151533/236615832-7fe453f8-a37b-481b-b64b-c2975ccaf6b2.png)

Q1: All check menus have position '-4'. However, they are separated by the Shell menus for the 'second menu'. Why?
Q2: when using
where=(window.is_tree and sel.count==0)
or
where=(window.parent.name=="NamespaceTreeControl" and sel.count==0 )
I can only choose the 'third menu'. However, how do I distinguish the 'first' from the 'second menu'?

Q3: command.refresh doesn't work for Navigation Pane. is there any way to refresh the Navigation Pane?

Q4: When creating a checked item, I have to use different commands for cmd=. How do I merge them? Something like cmd=IF(,,)? (Currently it works as I have created for each Item two new ones - one for checked=1 and one for checked=0 with different cmd=, but it looks terrible with so many repetitions)

Another strange thing I noticed on my computer. When I use the reg.set and reg.delete functions, sometimes they work, sometimes they don't!

$reg1 = 'HKEY_CLASSES_ROOT\...'
$reg2 = 'HKEY_CURRENT_USER\...'
$reg3 = 'HKEY_LOCAL_MACHINE\...'

reg.get(reg1,... - work
reg.set(reg1,... - work
reg.delete(reg1,... - work
reg.get(reg2,... - work
reg.set(reg2,... - work
reg.delete(reg2,... - work
reg.get(reg3,... - work
reg.set(reg3,... - DON'T work
reg.delete(reg3,... - DON'T work

With the ones I have used so far, I notice that when using HKLM the reg function cannot change values, regardless of whether they can be read.

Q5: Assuming the reg function works, why can't I change the values for HKLM?

**Comment by moudey** [comment]
>Q1: All check menus have position '-4'. However, they are separated by the Shell menus for the 'second menu'. Why?

Explain more.

>  Q2: when using
where=(window.is_tree and sel.count==0)
or
where=(window.parent.name=="NamespaceTreeControl" and sel.count==0 )
I can only choose the 'third menu'. However, how do I distinguish the 'first' from the 'second menu'?

It is determined by (the names of the menu items, menu items count, the names of the windows, the path of the selected objects, their type, etc.).

>Q3: command.refresh doesn't work for Navigation Pane. is there any way to refresh the Navigation Pane?

`command.refresh` works for me with the navigation pane. The implementation of the system's commands will be developed later and most of these commands will be included.

>Q4: When creating a checked item, I have to use different commands for cmd=. How do I merge them? Something like cmd=IF(,,)? (Currently it works as I have created for each Item two new ones - one for checked=1 and one for checked=0 with different cmd=, but it looks terrible with so many repetitions)

Give me a simple example to clarify the idea further so that a solution can be found.

> Q5: Assuming the reg function works, why can't I change the values for HKLM?

These commands need to run the explorer with administrator privileges.


Sorry, I've been getting really nervous lately and my focus has been distracted

**Comment by RubicBG** [comment]
First of all : Great Work!!!

I'm not a programmer, but with Nilesoft Shell I feel like a Microsoft programmer developing the new context menu for Windows 12! I am so excited!

It all started very simply: I just wanted to add an item, just one... now the menu looks like this:
![image](https://user-images.githubusercontent.com/127151533/236642707-82483815-eb2c-431c-ae91-bda885619c78.png)
With as many Items as I can add, it took me more than a day to position them:

Regarding Q1: when I use
modify(find='"Show this PC"' type='*' where=(window.is_tree) tip=this.pos)
to see how a standard (built-in) Item is positioned. I always get '-4' - for all items. 
![image](https://user-images.githubusercontent.com/127151533/236642737-a5567e07-83af-4ac6-acdd-2532a4dae670.png)
![image](https://user-images.githubusercontent.com/127151533/236642743-a36a1b84-ce50-4685-85a6-09d19f99e784.png)
Then how do the menus 'Terminal', 'File Manage', 'Develop' and 'Go to' come between items with the same position '-4'? If I give to standard Item a fixed position with pos='Number', they are positioned, but when I add the new items, again with pos='Number' they are arranged in a jumbled manner - they are not in the order in which I put them. I don't see any logic - I don't know what's going on, and I don't know how to explain it more precisely, and I don't know what to ask! (However I managed by positioning them all using pos='bottom' and then making the new items with position=indexof() ) If there is a logical explanation I want to know what it is.

Regarding Q2: I use
'type='*' where=(window.is_tree and sel.count==0)'
thus it works with no object selected.
I haven't been able to find a way where by right-clicking between two objects (there is a space between every two objects), I can add the same Items without them being displayed when the object is selected - these are two different menus (shown above). I wanted to find a solution myself and spent half a day wandering around with no results.

Regarding Q3: You gave me an answer here. just for information: I use
cmd=reg.set( .. ) & command.refresh
I see refreshing on the right in the View Pane, but on the left in the Navigation Pane nothing happens. If I press F5 the update happens.

Regarding Q4: This is the simplest code that applies to this question. All others are more complicated:



$regLinux = 'HKCU\Software\Classes\CLSID\{B2B4A4D1-2754-4140-A2EB-9A76D9D7CDC6}'
$regLinuxShowed = reg.get(regLinux,'System.IsPinnedToNameSpaceTree')!=0

item(type='*' where=(window.is_tree and sel.count==0 and regLinuxShowed)
	checked=regLinuxShowed title='Show Linux' position=indexof("Show Network", 1)
	cmd=reg.set(regLinux, 'System.IsPinnedToNameSpaceTree', '0', reg.dword)
	& command.refresh)

item(type='*' where=(window.is_tree and sel.count==0 and !regLinuxShowed) 
	checked=regLinuxShowed title='Show Linux' position=indexof("Show Network", 1)
	cmd=reg.delete(regLinux, 'System.IsPinnedToNameSpaceTree')
	& command.refresh)`



I can't merge commands under one cmd= . One item has reg.set and the other reg.delete (command.refresh doesn't work here either)

Regarding Q5: You gave me an answer here. I'll try using 'admin cmd args=' to force it.

**Comment by moudey** [comment]
> Then how do the menus 'Terminal', 'File Manage', 'Develop' and 'Go to' come between items with the same position '-4'? If I give to standard Item a fixed position with pos='Number', they are positioned, but when I add the new items, again with pos='Number' they are arranged in a jumbled manner - they are not in the order in which I put them. I don't see any logic - I don't know what's going on, and I don't know how to explain it more precisely, and I don't know what to ask! (However I managed by positioning them all using pos='bottom' and then making the new items with position=indexof() ) If there is a logical explanation I want to know what it is.

Position values
```
auto = -4
middle = -3
top = -2
bottom = - 1
custome >= 0
```
The "this.pos" function returns the item's default position value `auto`  unless it has been modified in the position property.
